### PR TITLE
Task1. Linear Regression and Gradient Descent

### PR DESCRIPTION
**Pull request by kirill-ivanov-a**
# Задача 1. Линейная регрессия, метод градиентного спуска

* **Дедлайн**: 31.10.2022, 23:59
* Основной полный балл: 5
* Максимум баллов: 11

## Задача

- [x] Самостоятельно реализовать функцию **gradient_descent(X, y)**, которая по заданной обучающей выборке обучает модель линейной регрессии, оптимизируя функционал методом градиентного спуска (Batch Gradient Descent, GD) и возвращая вектор весов **w**. В качестве функционала можно выбрать, например, функцию ошибок **MSE** + $L_2$-регуляризатор. Использовать матрично-векторные операции для вычисления градиента.
- [x] Найти данные, на которых интересно будет решать задачу регрессии. Зависимость целового признака от нецелевых должна быть не слишком сложной, чтобы обученная линейная модель смогла показать приемлимый результат. В крайнем случае взять данные для предсказания стоимости машин [тут](https://github.com/rustam-azimov/ml-course/tree/main/data/car_price) (целевой признак для предсказания --- **price**).
- [x] Считать данные, выполнить первичный анализ данных, при необходимости произвести чистку данных (Data Cleaning).
- [x] Выполнить разведочный анализ (EDA), использовать визуализацию, сделать выводы, которые могут быть полезны при дальнейшем решении задачи регрессии.
- [x] При необходимости выполнить полезные преобразования данных (например, трансформировать категариальные признаки в количественные), убрать ненужные признаки, создать новые (Feature Engineering).
- [x] Случайным образом разбить данные на обучающую и тестовую выборки, используя методы существующих библиотек.
- [x] При обучении моделей использовать масштабирование данных.
- [x] Обучить модель на обучающей выборке, используя функцию **gradient_descent(X, y)**. Оценить качество модели на обучающей и тестовой выборках, используя **MSE**, **RMSE** и $R^2$.
- [x] Обучить модель, используя существующую библиотеку. Например, в **sklearn** для $L_2$-регуляризатора можно использовать **Ridge**. Сравнить качество с вашей реализацией.
- [x] Повторить тоже самое, но используя кросс-валидацию.
- [x] Создать таблицу, со строками (mse-train, mse-test, rmse-train, rmse-test, r2-train, r2-test) и столбцами (Fold1, Fold2, ..., Foldk, E, STD), где k --- количество фолдов в кросс-валидации, E --- мат. ожидание и STD --- стандартное отклонение. Сделать выводы.
- [x] * (+1 балл) Перед обучением моделей подобрать наилучшее количество (и само подмножество) признаков, например используя Recursive Feature Elimination (RFE).
- [ ] * (+1 балл) Во все ваши реализации добавить возможность настройки нужных гиперпараметров, а в процессе обучения **всех** моделей осуществить подбор оптимальных значений этих гиперпараметров.
- [x] * (+2 балла) Также самостоятельно реализовать метод стохастического градиентного спуска (Stochastic Gradient Descent, SGD), обучить модели и добавить их во все сравнения.
- [x] * (+2 балла) Также самостоятельно реализовать метод мини-пакетного градиентного спуска (Mini Batch Gradient Descent), обучить модели и добавить их во все сравнения.

### CODE REVIEW

File: src/task01_linregr_gd/linear_regression/linear_regression.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+__all__ = ["LinearRegression"]
+
+
+class LinearRegression:
+    def __init__(
+        self,
+        descent: GradientDescentBase,
+        tolerance: float,
+        max_iter: int,
```
Comment (rustam-azimov) [inline src/task01_linregr_gd/linear_regression/linear_regression.py:18]: Можно попробовать побольше итераций дать, чем 200 у Вас в отчете. Другим методам хватило 200, но в стохастическом градиентном спуске итерации намного быстрее вычисляются, но их больше нужно.

**Review by rustam-azimov — APPROVED** [review]
10/11
